keps/sig-api-machinery/1929-built-in-default/README.md

```diff
@@ -0,0 +1,318 @@
+
+One very important nuance with CRDs is that CRDs are deserialized in
+unstructured objects that can distinguish between specified and
+unspecified fields. That is not the case for built-in types that are
+directly deserialized into Go-struct, where the unspecified fields are
+zero-valued. It's important to understand that only zero-value
+omit-empty fields will be defaulted using that mechanism.
+
+Computed defaults will not be supported as a declarative marker and will
+continue being supported through the existing defaulting functions.
+
+### Risks and Mitigations
+
+The main risk is that we could end-up with different defaults for
+existing types which would be extremely confusing for our users.
+
+## Design Details
+
+There are multiple challenges to running defaults declaratively and
+changing the current mechanisms:
+
+- We don't use OpenAPI to drive validation nor defaulting
+currently. There is no routing of the OpenAPI to most operations in
+kubernetes (the only thing right now is server-side apply).
+- OpenAPI fields don't strictly map to Go types fields
+(e.g. capitalization, embedding)
+- Reflection can't access comments, only structTags (which is why we have
+both structTags AND comments for SMP, but that's ugly and suboptimal).
+
+One possible way to address that second point is to serialize the object
+back to json, apply the default, and re-deserialize back into a
+go-struct, but that's memory intensive and Kubernetes doesn't need that.
+
+We're proposing to use the same mechanisms that we've put in place for
+server-side apply, implemented in
+"sigs.k8s.io/structured-merge-diff". We currently have a way for
+server-side apply to convert Go types into "json"-like values without
+serializing/deserializing and allocating memory by using reflection. We
+also have mechanisms in place to parse, process and use the OpenAPI.
+
+In other words, the plan would be the following:
+1. Update kube-openapi to parse the `+default` marker and include it in
+the OpenAPI definition
+2. Add additional "default" field to sigs.k8s.io/structured-merge-diff
+internal schema, and parse that field from the OpenAPI.
+3. Create a new "AddDefault" method in sigs.k8s.io/structured-merge-diff
+that receives a Value (value is an interface that can be backed by a
+reflection based Go-type, associated with the corresponding
+OpenAPI). That method would add the defaults in-place.
+4. Add and wire things so that this function can be called prior to
+running the existing defaulting code.
+
+### Marker format
+
+The marker will have the following format:
+```
+// +default=<any>
+```
+which describes the value that the server will use if the field has a
+zero-value.
+
+`<any>` has to be in one-line YAML format (including its json form),
+representing the actual value to use by default. That is different from
+the current encoding in kubebuilder, but there seems to be consensus
+that they would be able to change that to YAML too for that new marker.
```

Comment:
Namely: YAML preserves the concise semantics that you want for most values:

```
type Foo struct {
  // +default=32
  Integer int

  // +default=bar
  String string

  // +default=[popcorn, chips]
  StringList []string
}
```

while still allowing more complex expressions:

```
// +default=[{port: 80, name: http}]
```

Forcing the "one-line" subset allows us to sidestep a lot of the YAML weirdness in syntax, but we can still just use a YAML parser to parse the expressions if we want

Reply:
I just copied your great comment directly in there, thanks!